File: 743-network-delay-time/743-network-delay-time.cpp
using pii = pair<int,int>;
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        
        // form adjList //
        vector<vector<pii>> adjList(n+1);
        for(auto &x: times){
            
            int u = x[0], v = x[1], w = x[2];
            adjList[u].push_back({w, v});
            
        }
        
        // dijkstra //
        const int inf = 1e7;
        int dist[n+1];
        for(int i=0;i<n+1;++i)dist[i] = inf;
        
        dist[k] = 0;
        
        multiset<pii> ms; // {weight, node} //
            
        ms.insert({0, k});
        
        while(!ms.empty()){
            
            
            // extract the smallest dist node from the source node //
            pii top = *ms.begin();  
            ms.erase(ms.begin());
            int cur = top.second;
            
            // update the dist of all the adjacent nodes of this cur //
            
            for(pii adj: adjList[cur]){
                
                int weight = adj.first, adj_node = adj.second; 
                if( dist[cur] + weight < dist[adj_node] ){
                    int new_dist =  dist[cur] + weight;
                    dist[adj_node] = min(dist[adj_node], new_dist); 
                    ms.insert({ new_dist, adj_node }); 
                }
                
            }
            
            
        }
        
        
        
        int mx = 0;
        for(int i=1;i<=n;++i)mx = max(mx, dist[i]);
        
        if(mx == inf)return -1;
        return mx;
        
        
    }
};

File: 743-network-delay-time/743-network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        
        # Bellman-Ford # DP [node][k=jumps/hops]
        inf = 1e9
        graph = {}    
        
        # adding vertices to the graph #
        # for i in range(len(times)):
        #     if times[i][0] not in graph:
        #         graph[times[i][0]]=[]
        
        # adding edges #
        # for i in range(len(times)):
        #     graph[times[i][0]].append([ times[i][1], times[i][2] ])
        
        # return graph
        
#         _ = [None for _ in range(n)]
#         dp = [_ for x in range(n)]
        
        # dp = [  [inf for _ in range(n)] for x in range(n)]
        
        # return dp
        
        src = k
        # dp[src][0]=0
        dist = [inf for _ in range(n+1)]
        dist[src]=0
        for k in range(n+1):
            for e in times:
                u=e[0] 
                v=e[1]
                w=e[2]
                dist[v] = min(dist[v], w+dist[u])                
        
        time_taken = -1
        for i in range(1, len(dist)):
            time_taken=max(time_taken, dist[i])
        
        if time_taken>=inf:
            return -1
        else:
            return time_taken
        
        
        